numpy/core/src/multiarray/item_selection.c

```diff
@@ -2202,7 +2206,252 @@ count_nonzero_bytes(const npy_uint8 *d, npy_uintp unrollx)
+        target = MIN(target+innerloop_jump, end);
+        for (; d<target; d+=npyv_nlanes_u32) {
+            npyv_u32 vt = npyv_cvt_u32_b32(npyv_cmpeq_u32(npyv_load_u32(d), vzero)); 
+            vt = npyv_and_u32(vt, vone);
+            vsum32 = npyv_add_u32(vsum32, vt);
+        }
+        zero_count += npyv_sum_u32(vsum32);    
+    }
+
+    return unrollx - zero_count;
+}
+
+
+static NPY_INLINE NPY_GCC_OPT_3 npy_uintp
+count_nonzero_int64_simd(npy_int64 *d, npy_uintp unrollx)
+{
+    npy_uintp zero_count = 0;
+    const npy_int64 *end = d + unrollx;
+    const npyv_u64 vone = npyv_setall_u64(1); 
+    const npyv_u64 vzero = npyv_zero_u64();   
+    npyv_u64 vsum64 = npyv_zero_u64(); 
+
+    for (; d<end; d+=npyv_nlanes_u64) {
+        npyv_u64 vt = npyv_cvt_u64_b64(npyv_cmpeq_u64(npyv_load_u64(d), vzero)); 
+        vt = npyv_and_u64(vt, vone);
+        vsum64 = npyv_add_u64(vsum64, vt);
+    }
+
+    npy_uint64 sums[npyv_nlanes_u64];
+    npyv_store_u64(sums, vsum64);
+    for (int i=0; i<npyv_nlanes_u64; ++i) {
+        zero_count += sums[i];
+    }
+
+    return unrollx - zero_count;
+}
+
 #endif
+
+static NPY_INLINE NPY_GCC_OPT_3 npy_intp
+count_nonzero_int16(int ndim, const npy_int16 *data, const npy_intp *ashape, const npy_intp *astrides)
+{
+    int idim;
+    npy_intp shape[NPY_MAXDIMS], strides[NPY_MAXDIMS];
+    npy_intp coord[NPY_MAXDIMS];
+    npy_intp count = 0;
+    NPY_BEGIN_THREADS_DEF;
+
+    /* Use raw iteration with no heap memory allocation */
+    if (PyArray_PrepareOneRawArrayIter(
+                    ndim, ashape,
+                    data, astrides,
+                    &ndim, shape,
+                    &data, strides) < 0) {
+        return -1;
+    }
+
+    /* Handle zero-sized array */
+    if (shape[0] == 0) {
+        return 0;
+    }
+
+    NPY_BEGIN_THREADS_THRESHOLDED(shape[0]);
+    if (strides[0] == 2) {
+        NPY_RAW_ITER_START(idim, ndim, coord, shape) {
+            /* Process the innermost dimension */
+            const npy_int16 *d = data;
+            const npy_int16 *e = data + shape[0];
+            npy_uintp stride = shape[0] & -npyv_nlanes_u16;
+            count += count_nonzero_int16_simd(d, stride);
+            d += stride;
+            for (; d < e; ++d) {
+                count += (*d != 0);
+            }
+        } NPY_RAW_ITER_ONE_NEXT(idim, ndim, coord, shape, data, strides);
+    } else {
+        NPY_RAW_ITER_START(idim, ndim, coord, shape) {
+            npy_int16 *d = data;
+            /* Process the innermost dimension */
+            for (npy_intp i = 0; i < shape[0]; ++i, d = ((npy_int8*) d) + strides[0]) {
+                count += (*d != 0);
+            }
+        } NPY_RAW_ITER_ONE_NEXT(idim, ndim, coord, shape, data, strides);
+    }
+
+    NPY_END_THREADS;
+
+    return count;
+}
+
+
+static NPY_INLINE NPY_GCC_OPT_3 npy_intp
+count_nonzero_int32(int ndim, const npy_int32 *data, const npy_intp *ashape, const npy_intp *astrides)
```

Comment:
The similarities between `count_nonzero_int16/int32/int64` are too high, try to extract a public module or use the template syntax such as `einsum.c.src`.